File: src/main/java/top/mcfpp/nbt/tags/collection/ByteArrayTag.java
package top.mcfpp.nbt.tags.collection;

import org.apache.commons.lang3.ArrayUtils;
import top.mcfpp.nbt.tags.Tag;
import top.mcfpp.nbt.tags.TagType;
import top.mcfpp.nbt.tags.primitive.ByteTag;
import top.mcfpp.nbt.tags.primitive.NumericTag;
import top.mcfpp.nbt.visitors.StringTagVisitor;
import top.mcfpp.nbt.visitors.TagVisitor;

import java.util.Arrays;
import java.util.Optional;

public final class ByteArrayTag implements CollectionTag {
	private static final int SELF_SIZE_IN_BYTES = 24;
	public static final TagType<ByteArrayTag> TYPE = new TagType.VariableSize<ByteArrayTag>() {

		@Override
		public String getName() {
			return "BYTE[]";
		}

		@Override
		public String getPrettyName() {
			return "TAG_Byte_Array";
		}
	};
	private byte[] data;

	public ByteArrayTag(byte[] bs) {
		this.data = bs;
	}

	@Override
	public int sizeInBytes() {
		return SELF_SIZE_IN_BYTES + this.data.length;
	}

	@Override
	public byte getId() {
		return TAG_BYTE_ARRAY;
	}

	@Override
	public TagType<ByteArrayTag> getType() {
		return TYPE;
	}

	@Override
	public String toString() {
		StringTagVisitor stringTagVisitor = new StringTagVisitor();
		stringTagVisitor.visitByteArray(this);
		return stringTagVisitor.build();
	}

	@Override
	public Tag copy() {
		byte[] bs = new byte[this.data.length];
		System.arraycopy(this.data, 0, bs, 0, this.data.length);
		return new ByteArrayTag(bs);
	}

	public boolean equals(Object object) {
		return this == object || object instanceof ByteArrayTag && Arrays.equals(this.data, ((ByteArrayTag) object).data);
	}

	public int hashCode() {
		return Arrays.hashCode(this.data);
	}

	@Override
	public void accept(TagVisitor tagVisitor) {
		tagVisitor.visitByteArray(this);
	}

	public byte[] getAsByteArray() {
		return this.data;
	}

	@Override
	public int size() {
		return this.data.length;
	}

	public ByteTag get(int i) {
		return ByteTag.valueOf(this.data[i]);
	}

	@Override
	public boolean setTag(int i, Tag tag) {
		if (tag instanceof NumericTag numericTag) {
			this.data[i] = numericTag.byteValue();
			return true;
		} else {
			return false;
		}
	}

	@Override
	public boolean addTag(int i, Tag tag) {
		if (tag instanceof NumericTag numericTag) {
			this.data = ArrayUtils.add(this.data, i, numericTag.byteValue());
			return true;
		} else {
			return false;
		}
	}

	public ByteTag remove(int i) {
		byte b = this.data[i];
		this.data = ArrayUtils.remove(this.data, i);
		return ByteTag.valueOf(b);
	}

	@Override
	public void clear() {
		this.data = new byte[0];
	}

	@Override
	public Optional<byte[]> asByteArray() {
		return Optional.of(this.data);
	}

}

File: src/main/kotlin/top/mcfpp/nbt/tags/collection/ByteArrayTag.kt
package top.mcfpp.nbt.tags.collection

import org.apache.commons.lang3.ArrayUtils
import top.mcfpp.nbt.tags.Tag
import top.mcfpp.nbt.tags.primitive.ByteTag
import top.mcfpp.nbt.visitors.StringTagVisitor
import top.mcfpp.nbt.visitors.TagVisitor
import java.util.*

class ByteArrayTag(override var value: ByteArray) : CollectionTag<ByteTag, ByteArray> {
    override fun toString(): String {
        val stringTagVisitor = StringTagVisitor()
        stringTagVisitor.visitByteArray(this)
        return stringTagVisitor.build()
    }

    override fun copy(): ByteArrayTag {
        val bs = ByteArray(value.size)
        System.arraycopy(this.value, 0, bs, 0, value.size)
        return ByteArrayTag(bs)
    }

    override fun equals(other: Any?): Boolean {
        return this === other || other is ByteArrayTag && value.contentEquals(other.value)
    }

    override fun hashCode(): Int {
        return value.contentHashCode()
    }

    override fun accept(tagVisitor: TagVisitor) {
        tagVisitor.visitByteArray(this)
    }

    override val size: Int get() {
        return value.size
    }

    override operator fun get(index: Int): ByteTag {
        return ByteTag.valueOf(value[index])
    }

    override operator fun set(index: Int, tag: ByteTag): ByteTag {
        val b = value[index]
        value[index] = tag.byteValue()
        return ByteTag.valueOf(b)
    }

    override fun add(index: Int, tag: ByteTag) {
        this.value = ArrayUtils.insert(index, this.value, tag.byteValue())
    }

    override fun removeAt(index: Int): ByteTag {
        val b = value[index]
        this.value = ArrayUtils.remove(this.value, index)
        return ByteTag.valueOf(b)
    }

    override fun clear() {
        this.value = ByteArray(0)
    }

    override fun asByteArray(): ByteArray {
        return this.value
    }
}
